src/inference/particle_filter.jl

```diff
@@ -111,7 +111,16 @@ end
     particle_filter_step!(state::ParticleFilterState, new_args::Tuple, argdiffs,
         observations::ChoiceMap, proposal::GenerativeFunction, proposal_args::Tuple)
 
-Perform a particle filter update, where the model arguments are adjusted, new observations are added, and some combination of a custom proposal and the model's internal proposal is used for proposing new latent state (whatever is not proposed from the custom proposal will be proposed using the model's internal proposal).
+Perform a particle filter update, where the model arguments are adjusted, new
+observations are added, and some combination of a custom proposal and the
+model's internal proposal is used for proposing new latent state.  That is, for
+each particle, the value of a random choice at address `a` in the new trace for
+that particle is chosen by the following rule:
+* The particle's previous trace must not have a random choice at address `a`:
```

Comment:
@bzinberg Not sure about this phrasing: the new model trace includes both the old random choices *and* new random choices.

Reply:
> Come to think of it, I might be missing a piece of the picture. When would a user want to use `particle_filter_step!` with a proposal that is anything other than, essentially, the conditional distribution `P(x_{t+1} | x_t)`? Are there situations where the user specifies an HMM in some other way, or situations where the user specifies the HMM that way and gets inference benefits from using a different step proposal? (Rejuvenation proposals I get, but step proposals I'm not sure.) Would the main case be when the user gives a broader-than-prior proposal in order to increase particle diversity?

@bzinberg The optimal proposal would be P(x_{t + 1} | x_t, y_t) -- note the dependence on the newest observation y_t. But usually sampling (and evaluating the density of) the conditional distribution P(x_{t+1} | x_t, y_t) isn't possible. But it is possible to use step proposals that are better approximations to P(x_{t+1} | x_t, y_t) than P(x_{t + 1} | x_t) is.

Reply:
@binzberg I think a more precise characterization would be: "The random choices in the new trace satisfies the following conditions: (i) All addresses in the previous trace should also be in the new trace, with the same value, (ii) all newly introduced observed choices are included in the new trace, and (iii) all choices in the proposal choice map are present in the new trace. It is an error if no such trace exists in the support of the model (with the new arguments). If such a trace exists, then the random choices not determined by the previous trace, proposal choice map, or observations, are sampled using the internal proposal.

Reply:
Is there also some requirement to the effect of "the proposal cannot contain choices at addresses that exist in the old trace"?

Reply:
Yes. Technically, I think that's a consequence of (i) and (iii) because a choice cannot be taken from both the previous trace and the proposal choice map.

Reply:
It could if there was a requirement that it have the same value in both places.

Reply:
@marcoct, @alex-lew, updated the docstring based on your feedback -- please take a look when you get the chance